CMakeLists.txt

```diff
@@ -23,103 +23,128 @@ string(REGEX REPLACE ${VERSION_REGEX} "\\1" VERSION_STRING "${VERSION_STRING}")
+list(APPEND clang-tidy-fix "CLI11_CLANG_TIDY_EXE")
+list(APPEND clang-tidy-fix "CLI11_BUILD_EXAMPLES")
+
+list(APPEND build-docs "CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME")
+list(APPEND build-docs "CMAKE_VERSION VERSION_GREATER 3.10")
+list(APPEND build-docs "Doxygen_FOUND")
+
+option(CLI11_WARNINGS_AS_ERRORS "Turn all warnings into errors (for CI)")
+option(CLI11_SINGLE_FILE "Generate a single header file")
+
+cmake_dependent_option(CLI11_BUILD_DOCS
+  "Build CLI11 documentation" ON
+  "${build-docs}" OFF)
+
+cmake_dependent_option(CLI11_BUILD_TESTS
+  "Build CLI11 tests" ON
+  "BUILD_TESTING;CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME" OFF)
+
+cmake_dependent_option(CLI11_BUILD_EXAMPLES
+  "Build CLI11 examples" ON
+  "CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME" OFF)
+
+cmake_dependent_option(CLI11_BUILD_EXAMPLE_JSON
+  "Build CLI11 json example" OFF
+  "CLI11_BUILD_EXAMPLES" OFF)
+
+cmake_dependent_option(CLI11_SINGLE_FILE_TESTS
+  "Duplicate all the tests for a single file build" OFF
+  "BUILD_TESTING;CLI11_SINGLE_FILE" OFF)
+
+cmake_dependent_option(CLI11_INSTALL
+  "Install the CLI11 folder to include during install process" ON
+  "CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME" OFF)
+
+cmake_dependent_option(CLI11_CLANG_TIDY_FIX
+  "Perform fixes for clang-tidy" OFF
+  "${clang-tidy-fix}" OFF)
+
+cmake_dependent_option(CLI11_FORCE_LIBCXX
+  "Force clang to use libc++ instead of libstdc++ (Linux only)" OFF
+  "${force-libcxx}" OFF)
+
+if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME AND NOT DEFINED CMAKE_CXX_STANDARD)
+  set(CMAKE_CXX_STANDARD 11)
+endif()
+
+if (NOT DEFINED CMAKE_CXX_EXTENSIONS)
+  set(CMAKE_CXX_EXTENSIONS OFF)
+endif()
+
+if (NOT DEFINED CMAKE_CXX_STANDARD_REQUIRED)
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+endif()
+
+if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
+  set_property(GLOBAL PROPERTY USE_FOLDERS ON)
 else()
-    set(CUR_PROJ OFF)
+  get_property(cli11-use-folders GLOBAL PROPERTY USE_FOLDERS)
+  set_property(GLOBAL PROPERTY USE_FOLDERS ${cli11-use-folders})
 endif()
 
-# Allow dependent options
-include(CMakeDependentOption)
+if (CMAKE_VERSION VERSION_LESS 3.9)
+  message(STATUS "CMake 3.10 and later adds Doxygen support. Update CMake to build documentation")
+endif()
 
-# Allow IDE's to group targets into folders
-set_property(GLOBAL PROPERTY USE_FOLDERS ON)
+if (NOT Doxygen_FOUND)
+  message(STATUS "Doxygen not found, building docs has been disabled")
+endif()
 
-file(GLOB CLI11_headers "${CMAKE_CURRENT_SOURCE_DIR}/include/CLI/*")
-# To see in IDE, must be listed for target
+# Special target that adds warnings. Is not exported.
+add_library(CLI11_warnings INTERFACE)
 
+target_compile_options(CLI11_warnings
+  INTERFACE
+    $<$<BOOL:${CLI11_FORCE_LIBCXX}>:-stdlib=libc++>
+    $<$<CXX_COMPILER_ID:MSVC>:/W4 $<$<BOOL:${CLI11_WARNINGS_AS_ERRORS}>:/WX>>
+    $<$<NOT:$<CXX_COMPILER_ID:MSVC>>:-Wall -Wextra -pedantic -Wshadow $<$<BOOL:${CLI11_WARNINGS_AS_ERRORS}>:-Werror>>)
```

Comment:
Wouldn't an `$<IF...>` be a little nicer here?

Reply:
(I'm currently using this PR as an inspiration for some other CMake work I'm doing ;) )

Reply:
Oops, that was new in 3.8. Fine for what I'm doing, but not the current version of CLI11. :(

Reply:
Yeah unfortunately not having `$<IF:...>` has been a detriment to me fixing some other libraries up 😓, but it's still "doable". The generation engine handles it just fine, even if its not as fast.